bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java

```diff
@@ -0,0 +1,270 @@
+
+    private @Nullable ScheduledFuture<?> pollingJob;
+    private @Nullable ScheduledFuture<?> connectionMonitor;
+
+    private ZonedDateTime lastCommunication = ZonedDateTime.now();
+
+    // initialize timeBetweenDataShouldArrive with a large number
+    private Duration timeBetweenDataShouldArrive = Duration.ofDays(1);
+    private final Duration dataCanBeLateTolerance = Duration.ofSeconds(5);
+
+    // cached values for refresh command
+    private State statePM10 = UnDefType.UNDEF;
+    private State statePM25 = UnDefType.UNDEF;
+
+    public SDS011Handler(Thing thing, SerialPortManager serialPortManager) {
+        super(thing);
+        this.serialPortManager = serialPortManager;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // refresh channels with last received values from cache
+        if (RefreshType.REFRESH.equals(command)) {
+            if (NovaFineDustBindingConstants.CHANNEL_PM25.equals(channelUID.getId()) && statePM25 != UnDefType.UNDEF) {
+                updateState(NovaFineDustBindingConstants.CHANNEL_PM25, statePM25);
+            }
+            if (NovaFineDustBindingConstants.CHANNEL_PM10.equals(channelUID.getId()) && statePM10 != UnDefType.UNDEF) {
+                updateState(NovaFineDustBindingConstants.CHANNEL_PM10, statePM10);
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+
+        config = getConfigAs(NovaFineDustConfiguration.class);
+
+        if (!validateConfiguration()) {
+            return;
+        }
+
+        // parse ports and if the port is found, initialize the reader
+        SerialPortIdentifier portId = serialPortManager.getIdentifier(config.port);
+        if (portId == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, "Port is not known!");
+            return;
+        }
+
+        this.communicator = new SDS011Communicator(this, portId);
+
+        if (config.reporting) {
+            timeBetweenDataShouldArrive = Duration.ofMinutes(config.reportingInterval);
+            scheduler.schedule(() -> initializeCommunicator(WorkMode.REPORTING, timeBetweenDataShouldArrive), 0,
+                    TimeUnit.SECONDS);
+        } else {
+            timeBetweenDataShouldArrive = Duration.ofSeconds(config.pollingInterval);
+            scheduler.schedule(() -> initializeCommunicator(WorkMode.POLLING, timeBetweenDataShouldArrive), 0,
+                    TimeUnit.SECONDS);
+        }
+
+        Duration connectionMonitorStartDelay = timeBetweenDataShouldArrive.plus(CONNECTION_MONITOR_START_DELAY_OFFSET);
+        connectionMonitor = scheduler.scheduleWithFixedDelay(this::verifyIfStillConnected,
+                connectionMonitorStartDelay.getSeconds(), timeBetweenDataShouldArrive.getSeconds(), TimeUnit.SECONDS);
+    }
+
+    private void initializeCommunicator(WorkMode mode, Duration interval) {
```

Comment:
Just to be safe, I would make sure that `initialize`, `dispose`, and `initializeCommunicator` are synchronized.

Reply:
How should I do that? This won't work if I write `synchronized` on those methods.

The reason is that this will synchronize of the ´ThingHandler` object and this is never the same is a `dispose` and an `initialize` are called.  The `ThingManager` will always create a new `ThingHandler` instance before calling `initialize`, so the call to `dispose` will happen on a different (old) object.

This is what I wanted to state earlier: The openHAB framework currently does NOT ensure that `dispose` and `initialize` run into each other, if one is forced to schedule a task (that runs in a different thread) from `initialize` if one needs more than 5 seconds.

So either: I do NOT schedule this task and risk the warning "ThingHandler initialize takes longer than 5 seconds", or I take the (in my case pretty unlikely) risk that `dispose` and `initialize` collide eventually.

So which option should I pick?

Reply:
@J-N-K Could you please weigh in here? I can't tell if I'm being overly concerned or not with regards to thread safety here.

Reply:
I can chime in here. We never have `synchronized` keywords on such methods. As @t2000 mentions, the work is done in a separate thread and it is actually intended that the framework is able to dispose the handler, while such jobs might still be running. They should NOT block a call of `dispose`, so synchronizing it would be counter-productive.

Reply:
@kaikreuzer I'm just concerned that `initializeCommunicator` would be called after `dispose` since `initializeCommunicator` is called asynchronously.

Reply:
Sounds pretty unlikely and even if it does, the code should be written in a way that it shouldn't harm much.
If dispose would set the [communicator to null](https://github.com/openhab/openhab-addons/pull/7528/files/e685099b38e6dda40cf3326a1182d394983790ac#diff-ab4139e8f28a3d36d178fd52c9b9404bR213) and initializeCommunicator would [not try to set the Thing status](https://github.com/openhab/openhab-addons/pull/7528/files/e685099b38e6dda40cf3326a1182d394983790ac#diff-ab4139e8f28a3d36d178fd52c9b9404bR129-R133) upon a null-communicator, it should almost be safe ;-)